### ISSUE TITLE
Unable to insert string value into SQL_VARIANT column

### ISSUE THREAD

**Issue by gordthompson** [m32/sqlalchemy-tds#12]
SQLAlchemy: 2.0.32
sqlalchemy-tds: 1.0.1 - installed from current master 8a7f6ba
python-tds: 1.15.0

Attempting to insert a string value into a SQL_VARIANT column raises

> pytds.tds_base.OperationalError: ('Operand type clash: nvarchar(max) is incompatible with sql_variant', None)

```python
from sqlalchemy import Column, String, create_engine
from sqlalchemy.dialects.mssql import SQL_VARIANT
from sqlalchemy.orm import Session, declarative_base

connection_url = "mssql+pytds://scott:tiger^5HHH@192.168.0.199/test"
engine = create_engine(connection_url)
Base = declarative_base()

class Thing(Base):
    __tablename__ = "sqlvariant_test"
    val_type = Column(String(20), primary_key=True)
    val = Column(SQL_VARIANT, nullable=False)


Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

with Session(engine) as sess:
    sess.add(Thing(val_type="string", val="Hello World!"))
    sess.commit()
```
[traceback.txt](https://github.com/user-attachments/files/16750257/traceback.txt)

**Comment by gordthompson** [comment]
According to

https://learn.microsoft.com/en-us/sql/t-sql/data-types/sql-variant-transact-sql?view=sql-server-ver16

"The maximum length of the actual base type value is 8,000 bytes."

so string parameters for SQL_VARIANT columns could conceivably be capped at nvarchar(4000).

**Comment by m32** [comment]
You are right, unfortunately pytds does not support sql_variant type for type conversion.

When passing parameters to sp_execute pytds declares parameters and therefore their types, and the types are guessed from the types of python variables. Variables of type str are declared as @P1 NVARCHAR(MAX), and NVARCHAR(MAX) is not compatible with sql_variant.

I don't know how I can help you.

fails:
    sqlexec("INSERT INTO #sqlvariant_test values( %s, %s)", ('string', 'string'))
ok:
    sqlexec("INSERT INTO #sqlvariant_test values( 'integer', 'string')"

failing pytds only example:
```
import pytds
from t import username, userpass

conn = pytds.connect(dsn='127.0.0.1', user=username, password=userpass, database='testing')
def sqlexec(sql, params, fetch=False):
    c = conn.cursor()
    c.execute(sql, params)
    if fetch:
        for r in c.fetchall():
            print(r)
    c.close()
try:
    sqlexec('create table #sqlvariant_test(val_type varchar(20) primary key, val sql_variant)',())
    sqlexec("INSERT INTO #sqlvariant_test values( %s, %s)", ('integer', 1))
    sqlexec("INSERT INTO #sqlvariant_test values( %s, %s)", ('float', 1.1))
    sqlexec("INSERT INTO #sqlvariant_test values( %s, %s)", ('decimal(8,2)', 'CAST(1.1 as decimal(8,2))'))
    sqlexec("INSERT INTO #sqlvariant_test values( %s, %s)", ('string', 'string'))
    sqlexec("INSERT INTO #sqlvariant_test values( %s, %s)", ('string', b'bytes'))
    sqlexec("COMMIT", ())
    sqlexec("SELECT * FROM #sqlvariant_test", (), True)
finally:
    conn.close()

```

INFO:pytds:Sending query create table #sqlvariant_test(val_type varchar(20) primary key, val sql_variant)
DEBUG:pytds:Closing cursor
DEBUG:pytds:Closing cursor
make_param @P1 <sqltype:NVARCHAR(MAX)> 0 integer
make_param @P2 <sqltype:INT> 0 1
make_param  <sqltype:NVARCHAR(MAX)> 0 INSERT INTO #sqlvariant_test values( @P1, @P2)
make_param  <sqltype:NVARCHAR(MAX)> 0 @P1 NVARCHAR(MAX),@P2 INT
INFO:pytds:Sending RPC <pytds.tds_base.InternalProc object at 0x7f3d9a11b810> flags=0
DEBUG:pytds:Closing cursor
DEBUG:pytds:Closing cursor
make_param @P1 <sqltype:NVARCHAR(MAX)> 0 float
make_param @P2 <sqltype:FLOAT> 0 1.1
make_param  <sqltype:NVARCHAR(MAX)> 0 INSERT INTO #sqlvariant_test values( @P1, @P2)
make_param  <sqltype:NVARCHAR(MAX)> 0 @P1 NVARCHAR(MAX),@P2 FLOAT
INFO:pytds:Sending RPC <pytds.tds_base.InternalProc object at 0x7f3d9a11b810> flags=0
INFO:pytds:Sending CANCEL
DEBUG:pytds:Closing cursor
DEBUG:pytds:Closing cursor
make_param @P1 <sqltype:NVARCHAR(MAX)> 0 decimal(8,2)
make_param @P2 <sqltype:NVARCHAR(MAX)> 0 CAST(1.1 as decimal(8,2))
make_param  <sqltype:NVARCHAR(MAX)> 0 INSERT INTO #sqlvariant_test values( @P1, @P2)
make_param  <sqltype:NVARCHAR(MAX)> 0 @P1 NVARCHAR(MAX),@P2 NVARCHAR(MAX)
INFO:pytds:Sending RPC <pytds.tds_base.InternalProc object at 0x7f3d9a11b810> flags=0
INFO:pytds:Sending CANCEL
DEBUG:pytds:Closing connection
DEBUG:pytds:Closing all cursors which were opened by connection
DEBUG:pytds:Closing cursor
Traceback (most recent call last):
    sqlexec("INSERT INTO #sqlvariant_test values( %s, %s)", ('decimal(8,2)', 'CAST(1.1 as decimal(8,2))'))
    c.execute(sql, params)
    self._session.execute(operation, params)
    self.find_result_or_done()
    self.process_end(marker)
    self.raise_db_exception()
    raise ex
pytds.tds_base.OperationalError: ('Operand type clash: nvarchar(max) is incompatible with sql_variant', None)

**Comment by m32** [comment]
Is even worse, MS is inconsistent, can't declare parameter:
declare @p1 varchar(8000)
max value is 4000 :(

create table #sqlvariant_test(val_type varchar(20) primary key, val sql_variant)
declare @P1 NVARCHAR(4000) = 'integer'
declare @P2 INT = 1
INSERT INTO #sqlvariant_test values( @P1, @P2)

select @@VERSION

Microsoft SQL Server 2022 (RTM-CU12) (KB5033663) - 16.0.4115.5 (X64) Mar 4 2024 08:56:10 Copyright (C) 2022 Microsoft Corporation Developer Edition (64-bit) on Linux (Ubuntu 22.04.4 LTS) <X64>

**Comment by m32** [comment]
I reported it as:
https://github.com/denisenkom/pytds/issues/168

I can't do anything more

**Comment by gordthompson** [comment]
Okay, thanks for looking into it.

**Comment by m32** [comment]
I added explicit parameter conversion on insert/update for sql_variant type column.
Please check how the git version works now

example below
```
import decimal
from sqlalchemy import Column, String, create_engine, select, update
from sqlalchemy.dialects.mssql import SQL_VARIANT
from sqlalchemy.orm import Session, declarative_base

from t import username, userpass
engine = create_engine('mssql+pytds://'+username+':'+userpass.replace('@', '%40')+'@127.0.0.1/testing')
Base = declarative_base()

class Thing(Base):
    __tablename__ = "#sqlvariant_test"
    val_type = Column(String(20), primary_key=True)
    val = Column(SQL_VARIANT, nullable=True)


Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

with Session(engine) as sess:
    sess.add(Thing(val_type="py_int", val=1))
    sess.add(Thing(val_type="py_decimal", val=decimal.Decimal('1.1')))
    sess.add(Thing(val_type="py_float", val=1.2))
    # WTF TODO
    # after uncommenting the line below everything is ok
    #sess.commit()
    sess.add(Thing(val_type="py_string", val="Hello World!"))
    sess.commit()
    sess.add(Thing(val_type="py_bytes", val=b"Hello World!"))
    sess.add(Thing(val_type="py_none", val=None))
    sess.commit()

    stmt = select(Thing.val_type, Thing.val).order_by(Thing.val_type)
    rows = sess.execute(stmt).fetchall()
    for row in rows:
        print(row)
    sess.commit()

    # stmt = update(Thing).values(val=2.2).where(Thing.val_type=='py_float')
    stmt = update(Thing).values(val=b'ala ma kota').where(Thing.val_type=='py_bytes')
    print('stmt', str(stmt))
    sess.execute(stmt)
    sess.commit()

    stmt = select(Thing.val_type, Thing.val).order_by(Thing.val_type)
    rows = sess.execute(stmt).fetchall()
    for row in rows:
        print(row)
    sess.commit()

```

**Comment by gordthompson** [comment]
> Please check how the git version works now

Works fine for me after I uncomment the "WTF TODO" `.commit()`. :) Thanks for following-up on this!

**Comment by gordthompson** [comment]
I do notice one thing: When the code dumps the values at the end the "py_bytes" entry appears as 

```
('py_bytes', 'ala ma kota')
```

that is `'ala ma kota'` instead of `b'ala ma kota'`

**Comment by gordthompson** [comment]
Also, this fails:

```python
    stmt = update(Thing).values(val=b'\xab\xcd').where(Thing.val_type=='py_bytes')
```

```
sqlalchemy.exc.DatabaseError: (pytds.tds_base.DatabaseError) (UnicodeDecodeError('utf-8', b'\xab\xcd', 0, 1, 'invalid start byte'), None)
[SQL: UPDATE [#sqlvariant_test] SET val=%(val)s WHERE [#sqlvariant_test].val_type = %(val_type_1)s]
[parameters: {'val': <pytds.tds_base.Param object at 0x0000004591055C00>, 'val_type_1': 'py_bytes'}]
```

**Comment by m32** [comment]
I changed the type to varbinary and it's better, please check
    stmt = update(Thing).values(val=b'\xab\xcd'*(4000-1)).where(Thing.val_type=='py_bytes')

**Comment by gordthompson** [comment]
Looks good. I created a test:

https://github.com/gordthompson/sqlalchemy-tds/blob/sql_variant_test_wip/test/test_sql_variant.py

I can send you a PR if you'd like.

**Comment by m32** [comment]
Of course yes, thank you.

**Comment by m32** [comment]
I've published a new version 1.0.1 on pypi with fixes from this report. Thanks for your help and feedback.
